sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,446 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.jdbc;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import java.io.Serializable;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.sql.DataSource;
+
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>IO to read and write data on JDBC.</p>
+ * <p>
+ * <h3>Reading from JDBC datasource</h3>
+ * <p>
+ * JdbcIO source returns a bounded collection of {@code T} as a
+ * {@code PCollection<T>}. T is the type returned by the provided {@link RowMapper}.
+ * </p>
+ * <p>
+ * To configure the JDBC source, you have to provide a {@link DataSource}. The username and
+ * password to connect to the database are optionals. The following example illustrates how to
+ * configure a JDBC source:
+ * </p>
+ * <pre>
+ *   {@code
+ *
+ * pipeline.apply(JdbcIO.read()
+ *   .withDataSource(myDataSource)
+ *   .withStatement("select id,name from Person")
+ *   .withRowMapper(new JdbcIO.RowMapper<KV<Integer, String>>() {
+ *     public KV<Integer, String> mapRow(ResultSet resultSet) {
+ *       KV<Integer, String> kv = KV.of(resultSet.getInt(1), resultSet.getString(2));
+ *     }
+ *   })
+ *
+ *   }
+ * </pre>
+ * <p>
+ * You can find a full {@link RowMapper} example in {@code JdbcIOTest}.
```

Comment:
The example in JdbcIOTest seems not particularly more complete than the one in this doc; maybe it's not worth mentioning here. Likewise for the write.